src/s2/s2predicates.cc

```diff
@@ -789,9 +789,256 @@ int CompareEdgePairDistance(const S2Point& a0, const S2Point& a1,
+  // the gappa proof at end of the file).
+  //
+  // NOTE: If we want to push this error bound down as far as possible, we could
+  // use the dot product algorithm created by Ogita et al:
+  //
+  //   Accurate Sum and Dot Product, Ogita, Rump, Oishi 2005.
+  //
+  // Along with the 2x2 determinant algorithm by Kahan (which is useful for any
+  // bilinear form):
+  //
+  //   Further Analysis of Kahan's Algorithm for the Accurate Computation of
+  //   2x2 Determinants, Jeannerod, Louvet, and Muller, 2013.
+  //
+  // Both algorithms allow us to have bounded relative error, and since we're
+  // only interested in the sign of this operation, as long as the relative
+  // error is < 1 we can never get a sign flip, which would make this exact for
+  // our purposes.
+  constexpr T kMaxError = 32 * std::numeric_limits<T>::epsilon();
+
+  T mdota = m.DotProd(a);
+  T mdotb = m.DotProd(b);
+  T mdotc = m.DotProd(c);
+  T mdotd = m.DotProd(d);
+
+  T ndota = n.DotProd(a);
+  T ndotb = n.DotProd(b);
+  T ndotc = n.DotProd(c);
+  T ndotd = n.DotProd(d);
+
+  T prodab = mdota * ndotb - mdotb * ndota;
+  T prodcd = mdotc * ndotd - mdotd * ndotc;
+
+  if (std::fabs(prodab - prodcd) > kMaxError) {
+    return prodab < prodcd ? -1 : +1;
+  }
+  return 0;
+}
+
+int ExactIntersectionOrdering(const Vector3_xf& a, const Vector3_xf& b,
+                              const Vector3_xf& c, const Vector3_xf& d,
+                              const Vector3_xf& m, const Vector3_xf& n) {
+  S2_DCHECK_NE(a, b);
+  S2_DCHECK_NE(a, -b);
+  S2_DCHECK_NE(c, d);
+  S2_DCHECK_NE(c, -d);
+  S2_DCHECK_NE(n, m);
+  S2_DCHECK_NE(n, -m);
+
+  ExactFloat mdota = m.DotProd(a);
+  ExactFloat mdotb = m.DotProd(b);
+  ExactFloat mdotc = m.DotProd(c);
+  ExactFloat mdotd = m.DotProd(d);
+
+  ExactFloat ndota = n.DotProd(a);
+  ExactFloat ndotb = n.DotProd(b);
+  ExactFloat ndotc = n.DotProd(c);
+  ExactFloat ndotd = n.DotProd(d);
+
+  ExactFloat prodab = mdota * ndotb - mdotb * ndota;
+  ExactFloat prodcd = mdotc * ndotd - mdotd * ndotc;
+
+  if (prodab < prodcd) return -1;
+  if (prodab > prodcd) return +1;
+  return 0;
+}
+
+int CircleEdgeIntersectionOrdering(const S2Point& a, const S2Point& b,
+                                   const S2Point& c, const S2Point& d,
+                                   const S2Point& m, const S2Point& n) {
+  int ans = TriageIntersectionOrdering(a, b, c, d, m, n);
+  if (ans != 0) {
+    return ans;
+  }
+
+  // We got zero, check for duplicate/reverse duplicate edges before falling
+  // back to more precision.
+  if ((a == c && b == d) || (a == d && b == c)) {
+    return 0;
+  }
+
+  if (kHasLongDouble) {
+    ans = TriageIntersectionOrdering(ToLD(a), ToLD(b), ToLD(c), ToLD(d),
+                                     ToLD(m), ToLD(n));
+    if (ans != 0) {
+      return ans;
+    }
+  }
+
+  return ExactIntersectionOrdering(  //
+      ToExact(a), ToExact(b), ToExact(c), ToExact(d), ToExact(m), ToExact(n));
+
+  return 0;
```

Comment:
There two returns here (line 913 && 916)

Reply:
Shocked the linter doesn't catch this, I'll get it fixed for the next roll.  Glad to see you're still contributing btw =D